orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton.go

```diff
@@ -0,0 +1,185 @@
+}
+
+// TODO: to be setup and used in M2 Part C
+const reindexLoopInterval = 5 * time.Second
+
+func NewReindexerSingleton(store Store, versioner Versioner) Reindexer {
+	return &reindexerSingleton{store: store, versioner: versioner}
+}
+
+func (r *reindexerSingleton) Run(ctx context.Context) {
+	const indexerID = ""
+	var sendUpdate func(string) = nil
+
+	batches := r.getReindexBatches(ctx)
+	for {
+		if isCanceled(ctx) {
+			glog.Warning("State reindexing async job canceled")
+			return
+		}
+
+		err := r.reindexJobs(ctx, indexerID, batches, sendUpdate)
+
+		// NOTE: consider sleeping when no err for perf reasons
+		if err != nil {
+			clock.Sleep(failedJobSleep)
+		} else {
+			clock.Sleep(reindexLoopInterval)
+		}
+	}
+}
+
+func (r *reindexerSingleton) RunUnsafe(ctx context.Context, indexerID string, sendUpdate func(string)) error {
+	batches := r.getReindexBatches(ctx)
+	return r.reindexJobs(ctx, indexerID, batches, sendUpdate)
+}
+
+func (r *reindexerSingleton) GetIndexerVersions() ([]*indexer.Versions, error) {
+	return r.versioner.GetIndexerVersions()
+}
+
+// Get network-segregated reindex batches with capped number of state IDs per batch.
+func (r *reindexerSingleton) getReindexBatches(ctx context.Context) []reindexBatch {
+	var idsByNetwork state_types.IDsByNetwork
+	for {
+		if isCanceled(ctx) {
+			return nil
+		}
+		ids, err := r.store.GetAllIDs()
+		if err != nil {
+			err = wrap(err, ErrDefault, "")
+			glog.Errorf("Failed to get all state IDs for state indexer reindexing, will retry: %v", err)
+			clock.Sleep(failedGetBatchesSleep)
+			continue
+		}
+		idsByNetwork = ids
+		break
+	}
+
+	var current, rest state_types.IDs
+	var batches []reindexBatch
+	for networkID, ids := range idsByNetwork {
+		rest = ids
+		for len(rest) > 0 {
+			count := util.MinInt(numStatesToReindexPerCall, len(rest))
+			current, rest = rest[:count], rest[count:]
+			batches = append(batches, reindexBatch{networkID: networkID, stateIDs: current})
+		}
+	}
+	return batches
+}
+
+func (r *reindexerSingleton) reindexJobs(ctx context.Context, indexerID string, batches []reindexBatch, sendUpdate func(string)) error {
+	jobs, err := r.getJobs(indexerID)
+	if err != nil {
+		return wrap(err, ErrDefault, indexerID)
+	}
+	// QUESTION: why return error if no jobs? this is brought over from Run / RunUnsafe of reindexer_queue
+	if len(jobs) == 0 {
+		return merrors.ErrNotFound
+	}
+
+	for _, j := range jobs {
+		err = r.reindexJob(j, ctx, batches, sendUpdate)
+		if err != nil {
+			return wrap(err, ErrDefault, indexerID)
+		}
+	}
+	return nil
+}
+
+func (r *reindexerSingleton) reindexJob(job *Job, ctx context.Context, batches []reindexBatch, sendUpdate func(string)) error {
+	defer TestHookReindexDone()
+
+	start := clock.Now()
+
+	jobErr := executeJob(ctx, job, batches)
+
+	// TODO(reginawang3495) Add indexer fail count increase and logging
+	// Set Indexer Version every time for now
+	err := r.versioner.SetIndexerActualVersion(job.Idx.GetID(), job.To)
```

Comment:
Only set actual version if there's no jobErr

Reply:
For the sake of not redoing this infinite number of times, I purposefully set actual version even if `jobErr`. 

This is effectively making job retry count = 1. This will be modified with the TODO in the next PR

Thoughts?

Reply:
Hm. Correct me if I'm wrong, but this introduces a bug -- yes we won't retry at all, but also it will incorrectly mark jobs as completed even when they're not (?)

If that's the case, then I think we shouldn't commit the code as-is

Reply:
Yes it'll mark the jobs as completed even when they aren't -- Last time we had a sync up, I remember you said that this (marking jobs as complete) is the expected behavior once the `numTimesRetries >= maxRetries`. In this case, I am simply pretending that `maxRetries == 0` 

If this isn't correct, then I can change this

Reply:
Ah right. So difference between "job complete" and "job successful"

Since we don't have a job queue anymore, we don't need to distinguish. So we *don't* want to update an indexer's version unless it was successfully reindexed

I think actually, we can go ahead and skip that concept we talked about of tracking numRetries -- it's fine if we endlessly retry to reindex a broken indexer, we just need to add a long sleep between attempts